app/webpacker/controllers/bulk_form_controller.js

```diff
@@ -0,0 +1,21 @@
+import { Controller } from "stimulus";
+
+// Manages "modified" state for a form with multiple records
+export default class BulkFormController extends Controller {
+  connect() {
+    this.form = this.element;
+
+    // Start listening for any changes within the form
+    // this.element.addEventListener('change', this.toggleModified.bind(this)); // dunno why this doesn't work
+    for (const element of this.form.elements) {
+      element.addEventListener("keyup", this.toggleModified.bind(this)); // instant response
+      element.addEventListener("change", this.toggleModified.bind(this)); // just in case (eg right-click paste)
+    }
+  }
+
+  toggleModified(e) {
+    const element = e.target;
+    const changed = element.value != element.defaultValue;
```

Comment:
Oh, cool, I didn't know about `defaultValue`. I thought we would need to read the `value` property of the element but I guess that the effect is exactly the same?

Reply:
The `value` of an element is then updated to its new value ; here we need to compare the current value with the _initial_ value, and see if it has changed, not only compared to its previous value.

Reply:
Yes, that is the value "attribute" in the HTML is available in JS as `.initialValue`. The current value "property" is available as `.value` of course.

Confusingly, the Javascript method `getAttribute("value")` actually returns the property value, not the initial HTML attribute value.

Reply:
To confuse things further, I think we'll need to also consider the value stored in the database, because there's one case where we reload the form with unsaved user input: when the record update has failed due to error. I'll investigate in the next PR...

Reply:
That's where lock numbers can be handy...